[SimplifyCFG] Avoid increasing too many phi entries when removing empty blocks

Chengjunp opened the pull request:
Now in the simplifycfg and jumpthreading passes, we will remove the empty blocks (blocks only have phis and an unconditional branch). However, in some cases, this will increase size of the IR and slow down the compile of other passes dramatically. For example, we have the following CFG:

1. BB1 has 100 predecessors, and unconditionally branches to BB2 (does not have any other instructions).
2. BB2 has 100 phis.

Then in this case, if we remove BB1, for every phi in BB2, we need to increase 99 entries (replace the incoming edge from BB1 with 100 edges from its predecessors). Then in total, we will increase 9900 phi entries, which can slow down the compile time for many other passes.

Therefore, in this change, we add a check to see whether removing the empty blocks will increase lots of phi entries. Now, the threshold is 1000 (can be controlled by the command line option `max-phi-entries-increase-after-removing-empty-block`), which means that we will not remove an empty block if it will increase the total number of phi entries by 1000. This threshold is conservative and for most of the cases, we will not have such a large phi. So, this will only be triggered in some unusual IRs.

Chengjunp committed:
Avoid introducing complex phis after removing empty blocks in the cfg

nikic commented on llvm/lib/Transforms/Utils/Local.cpp:
```diff
@@ -1040,6 +1045,42 @@ CanRedirectPredsOfEmptyBBToSucc(BasicBlock *BB, BasicBlock *Succ,
   return true;
 }
 
+// Check whether removing BB will make the phis in its Succ will have too
```
```suggestion
// Check whether removing BB will make the phis in its Succ have too
```

nikic commented on llvm/lib/Transforms/Utils/Local.cpp:
```diff
@@ -1040,6 +1045,42 @@ CanRedirectPredsOfEmptyBBToSucc(BasicBlock *BB, BasicBlock *Succ,
+  int NumPreds = pred_size(BB);
+  auto *Succ = BB->getTerminator()->getSuccessor(0);
+  for (auto &Phi : Succ->phis()) {
+    auto BlockIdx = Phi.getBasicBlockIndex(BB);
```
```suggestion
    int BlockIdx = Phi.getBasicBlockIndex(BB);
```

nikic commented on llvm/lib/Transforms/Utils/Local.cpp:
```diff
@@ -1040,6 +1045,42 @@ CanRedirectPredsOfEmptyBBToSucc(BasicBlock *BB, BasicBlock *Succ,
+static bool introduceTooComplexPhi(BasicBlock *BB) {
+  // Check BB only has phi and an unconditional branch
+  BranchInst *Branch = dyn_cast<BranchInst>(BB->getFirstNonPHIOrDbg(true));
+  assert(Branch && Branch->isUnconditional() && "BB is not an empty block");
```
Branch is unused outside the assertion, and getFirstNonPHIOrDbg() may be expensive.

I think it would be better to pass in BasicBlock *Succ to this function, like many other parts of this transform. The fact that it's an empty block with single successor is verified elsewhere.

Chengjunp replied:
I see. Thank you for pointing this out. I will update this.

nikic reviewed (commented):
The general direction of the change makes sense to me.

I'm slightly worried that this may cause regressions for switch transforms, where it may be normal to have very large phis.

A possible alternative heuristic would not look at the number of entries in one phi, but the total number of entries across all phis in the successor, so that we would limit less for one phi and more for a large number of phis. But hard to say in abstract whether that's better...

nikic commented:
Compile-time: https://llvm-compile-time-tracker.com/compare.php?from=bf68e9047f62c22ca87f9a4a7c59a46b3de06abb&to=c2b97e008d826915ca2d23697e00dec989dd72b1&stat=instructions:u Minor improvement on sqlite3 correlated with code size reduction.

Chengjunp committed:
Update llvm/lib/Transforms/Utils/Local.cpp

Co-authored-by: Nikita Popov <github@npopov.com>

Chengjunp committed:
Update llvm/lib/Transforms/Utils/Local.cpp

Co-authored-by: Nikita Popov <github@npopov.com>

Chengjunp commented:
Hi, thank you for the review. I think the minor improvement can be expected because this kind of pattern is unusual. I did encounter a very extreme case that by adding this check, it can save more than 20% compile time. 

In term of the heuristic, looking at the total number of entries across all Phis in the successor also makes sense to me. But I'm considering how to determine the threshold in that case. Should the threshold be the same regardless of the number of Phis in the successor?  Or the threshold should be linear to the number of Phis in the Successor?

dtcxzyw commented on llvm/lib/Transforms/Utils/Local.cpp:
```diff
@@ -1040,6 +1045,42 @@ CanRedirectPredsOfEmptyBBToSucc(BasicBlock *BB, BasicBlock *Succ,
+  auto *Succ = BB->getTerminator()->getSuccessor(0);
+  for (auto &Phi : Succ->phis()) {
+    int BlockIdx = Phi.getBasicBlockIndex(BB);
+    if (BlockIdx >= 0) {
```
`BlockIdx >= 0` is a loop invariant.

Chengjunp committed:
Update introduceTooComplexPhi

Chengjunp commented:
Hi @nikic, @dtcxzyw, I have updated the code. Could you help to have another review? Thanks!

nikic commented:
> In term of the heuristic, looking at the total number of entries across all Phis in the successor also makes sense to me. But I'm considering how to determine the threshold in that case. Should the threshold be the same regardless of the number of Phis in the successor? Or the threshold should be linear to the number of Phis in the Successor?

If I'm understanding correctly, what we really want to prevent is a quadratic increase in phi entries, as we may be adding up to `#phis * #preds` new entries. So I think that is the quantity we should be limiting with a fixed limit (though it should be higher than 100 in that case -- would a limit of 1000 work?)

XChy commented:
Good catch. Seem to be related to slowing down in #86312.

Chengjunp committed:
Change heuristic to check the increase in # phi entries

Chengjunp commented:
Hi @nikic, @dtcxzyw, @XChy, I have updated the heuristic so that it will look at the increase of the number phi entries to determine whether removing a block or not.
But this change has one issue that if there are many empty blocks in the IR and removing any of them will not increase the number of phi entries greatly but removing all of them will have a large increase in the total phi entries. For this case, currently heuristic will remove all of them. 
Therefore, I'm wondering whether we should have something like a global map in the simplifycfg to track how many new phi entries we have already introduced to each block and stop further removing the blocks if the total new phi entries in one block reach a threshold (for example, 1000).

Chengjunp commented:
Hi @nikic, I took any look at the code, and I think it could be a little bit complex to track the total number of new phi entries of different blocks in a pass since the function `TryToSimplifyUncondBranchFromEmptyBlock` could be used in different passes. Could you help to review the current change and do a compile time test on it? Thank you very much!

Chengjunp committed:
Merge branch 'main' into chengjunp/AvoidComplexPhi

nikic commented:
> Hi @nikic, @dtcxzyw, @XChy, I have updated the heuristic so that it will look at the increase of the number phi entries to determine whether removing a block or not. But this change has one issue that if there are many empty blocks in the IR and removing any of them will not increase the number of phi entries greatly but removing all of them will have a large increase in the total phi entries. For this case, currently heuristic will remove all of them.

Does this patch still cover your original motivating case, or is the new version not sufficient due to this problem?

nikic commented on llvm/lib/Transforms/Utils/Local.cpp:
```diff
@@ -1047,6 +1053,34 @@ CanRedirectPredsOfEmptyBBToSucc(BasicBlock *BB, BasicBlock *Succ,
+  for (auto &Phi : Succ->phis()) {
+    // If the incoming value is a phi and the phi is defined in BB,
+    // then removing BB will not increase the total phi entries of the ir.
+    if (PHINode *IncomingPhi =
```
```suggestion
    if (auto *IncomingPhi =
```
Nit: Use auto when using dyn_cast.

Chengjunp committed:
Fix code style issues and update tests

Chengjunp commented:
> Does this patch still cover your original motivating case, or is the new version not sufficient due to this problem?

Sorry for the late reply. I have done the tests during the last few days and the current version can help improve the compile time in my original case. If everything looks fine, could you help review and merge this change? Thanks!

Chengjunp committed:
Format

dtcxzyw reviewed (approved):
LGTM. Thank you!
